src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs

```diff
@@ -135,93 +135,174 @@ private void LoadTestsFromAnExtension(string extensionAssembly, IEnumerable<stri
 
             foreach (var discoverer in discovererToSourcesMap.Keys)
             {
-                Type discovererType = null;
+                this.DiscoverTestsFromSingleDiscoverer(discoverer, discovererToSourcesMap, context, discoverySink, logger, ref discoverersFromDeprecatedLocations, ref totalAdaptersUsed, ref totalTimeTakenByAdapters);
+            }
 
-                // See if discoverer can be instantiated successfully else move next.
-                try
-                {
-                    discovererType = discoverer.Value.GetType();
-                }
-                catch (Exception e)
-                {
-                    var mesage = string.Format(
-                        CultureInfo.CurrentUICulture,
-                        CrossPlatEngineResources.DiscovererInstantiationException,
-                        e.Message);
-                    logger.SendMessage(TestMessageLevel.Warning, mesage);
-                    EqtTrace.Error("DiscovererEnumerator.LoadTestsFromAnExtension: {0} ", e);
+            if (this.discoveryResultCache.TotalDiscoveredTests == 0)
+            {
+                DiscovererEnumerator.LogWarningOnNoTestsDiscovered(sources, testCaseFilter, logger);
+            }
 
-                    continue;
-                }
+            this.CollectTelemetryAtEnd(totalTimeTakenByAdapters, totalAdaptersUsed);
+        }
+
+        private void CollectTelemetryAtEnd(double totalTimeTakenByAdapters, double totalAdaptersUsed)
+        {
+            // Collecting Total Time Taken by Adapters
+            this.requestData.MetricsCollection.Add(TelemetryDataConstants.TimeTakenInSecByAllAdapters,
+                totalTimeTakenByAdapters);
+
+            // Collecting Total Adapters Used to Discover tests
+            this.requestData.MetricsCollection.Add(TelemetryDataConstants.NumberOfAdapterUsedToDiscoverTests,
+                totalAdaptersUsed);
+        }
+
+        private void DiscoverTestsFromSingleDiscoverer(
+            LazyExtension<ITestDiscoverer, ITestDiscovererCapabilities> discoverer,
+            Dictionary<LazyExtension<ITestDiscoverer, ITestDiscovererCapabilities>,IEnumerable<string>> discovererToSourcesMap,
+            DiscoveryContext context,
+            TestCaseDiscoverySink discoverySink,
+            IMessageLogger logger,
+            ref bool discoverersFromDeprecatedLocations,
+            ref double totalAdaptersUsed,
+            ref double totalTimeTakenByAdapters)
+        {
+            if (TryToLoadDiscoverer(discoverer, logger, out var discovererType) == false)
```

Comment:
TryToLoadDiscoverer already returns bool. No need for == false.

Reply:
`!` is difficult to read. That's why kept false explicitly.

Reply:
if(!succuess) is more readable than if(success == false). We can have a second opinion here as well to know what is the recommended approach here.